File: 202410/23.merge-k-sorted-lists.cpp
/*
 * @lc app=leetcode id=23 lang=cpp
 *
 * [23] Merge k Sorted Lists
 */

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

struct MyListNode
{
    int val;
    MyListNode *next;
    MyListNode() : val(0), next(nullptr) {}
    MyListNode(int x) : val(x), next(nullptr) {}
    MyListNode(int x, MyListNode *next) : val(x), next(next) {}
};

// @lc code=start
class Solution
{
public:
    MyListNode *mergeKLists(vector<MyListNode *> &lists)
    {
        if (lists.size() == 0)
        {
            return nullptr;
        }

        MyListNode *dummy = new MyListNode();
        MyListNode *curr = dummy;
        auto cmp = [](MyListNode *a, MyListNode *b)
        { return a->val > b->val; };
        priority_queue<MyListNode *, vector<MyListNode *>, decltype(cmp)> pq(cmp);

        for (MyListNode *list : lists)
        {
            if (list != nullptr)
            {
                pq.push(list);
            }
        }

        while (!pq.empty())
        {
            MyListNode *top = pq.top();
            pq.pop();
            curr->next = top;
            curr = curr->next;
            if (top->next != nullptr)
            {
                pq.push(top->next);
            }
        }

        return dummy->next;
    }
};
// @lc code=end

File: 202503/23.Merge k Sorted Lists.java
/*
 * @lc app=leetcode id=23 lang=java
 * @lcpr version=30103
 *
 * [23] Merge k Sorted Lists
 */

import java.util.PriorityQueue;

// @lc code=start
class ListNode {

    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {

    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((l1, l2) -> l1.val - l2.val);
        int n = lists.length;
        int count = 0;

        ListNode dummy = new ListNode();
        ListNode p = dummy;

        for (int i = 0; i < n; i++) {
            ListNode list = lists[i];
            if (list != null) {
                pq.offer(list);
                lists[i] = list.next;
            }
        }

        while (!pq.isEmpty()) {
            ListNode node = pq.poll();
            p.next = node;

            if (node.next != null) {
                pq.offer(node.next);
            }
            p = p.next;
        }

        return dummy.next;
    }
}
// @lc code=end

/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

 */
